### PR TITLE
Add HTML validator action

### PR DESCRIPTION
**Pull request by krassowski**
Adds four CI jobs:
- build suite using Jekyll
- validate HTML
- check links (excluding absolute `/` links and LinkedIn links for now due to limitations of checker)
- run Lighthouse audits:
  - those get uploaded to temporary public storage on Google servers and can be accessed easily from the browser for 7 days

### CODE REVIEW

**Comment by krassowski** [comment]
The failures are intentional. I propose to merge and iterate by fixing the detected issues in subsequent PRs.

Should we set a threshold for lighthouse score?

File: .github/workflows/validate.yml
Comment (choldgraf) [inline .github/workflows/validate.yml:25]: We have `nox` usable with this repository now, so we could just re-use that here via a verb like `nox -s build`. That might reduce some redundancy.

https://github.com/jupyter/jupyter.github.io/blob/10cdc9296f64b5607b4f7e5ab46c21d31a25e7c9/noxfile.py#L15-L18

We'd need to update that so that there was a build verb that didn't also do `liveserve` but that would be easy

**Review by choldgraf — COMMENTED** [review]
This all sounds good to me, I think it's a great idea to add in some basic checks for reasonableness

File: noxfile.py
Comment (choldgraf) [inline noxfile.py:16]: how about we call this `build-live`

you can also use hyphens in the name by following a pattern like this:

https://github.com/2i2c-org/infrastructure/blob/b910e65544d35b354122b264037bc5a9fd6e1b61/noxfile.py#L17-L19

File: noxfile.py
Comment (choldgraf) [inline noxfile.py:21]: and we call this `build`

**Review by choldgraf — COMMENTED** [review]
quick idea on naming!

File: noxfile.py
Comment (krassowski) [reply noxfile.py:16]: Makes sense and thanks for the link! I though about doing it this way but was not sure if it was fine to change it since it is also used in the readme.

**Comment by krassowski** [comment]
So I was enthusiastic about using common nox command here at first, but I am no longer convinced. The old docker-based approach was building in 51 seconds, the nox/conda takes 6 minutes 24 seconds. It is a substantial delay. I can go for mamba, cache etc but it will all increase complexity and maintenance burden - caching is hard.

**Comment by choldgraf** [comment]
Huh, that's surprising. But if it's that big a difference I agree. We could probably optimize the nox build down but that's not the best use of time in this PR in my opinion, unless you're enthusiastic to try it out

**Comment by krassowski** [comment]
I guess it comes down to having the docker image already available on GitHub infrastructure (the docker image is the one from the action suggested by GitHub) vs having to resolve dependencies and install them from conda - I would not expect a big difference in the actual build step.

**Comment by choldgraf** [comment]
Yeah for sure - maybe with mamba it'd be significantly faster? This also seems like it'd be a one time thing I'd we cached it properly. Environments all get stored in a .nox folder so maybe we could just cache that?

**Comment by krassowski** [comment]
Can we address the install via nox/mamba/caching in a follow-up PR? Automation is exhausting ;)

**Comment by choldgraf** [comment]
Totally, I'm just thinking out loud

**Comment by krassowski** [comment]
Currently the lighthouse result show up as:

>   <details>
>
>   Total size of all the files uploaded is 1165820 bytes
>   Finished uploading artifact lighthouse-results. Reported size is 1165820 bytes. There were 0 items that failed to upload
>   Uploading median LHR of http://localhost:45821/about.html...success!
>   Open the report at https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1640813497053-27107.report.html
>   Uploading median LHR of http://localhost:45821/binder.html...success!
>   Open the report at https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1640813497456-16487.report.html
>   Uploading median LHR of http://localhost:45821/community.html...success!
>   Open the report at https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1640813497852-78305.report.html
>   Uploading median LHR of http://localhost:45821/documentation.html...success!
>   Open the report at https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1640813498306-7357.report.html
>   Uploading median LHR of http://localhost:45821/hub.html...success!
>   Open the report at https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1640813498846-65386.report.html
>   No GitHub token set, skipping GitHub status check.
>   Dumping 5 reports to disk at /home/runner/work/jupyter.github.io/jupyter.github.io/.lighthouseci...
>   Done writing reports to disk.
>  </details>
> 
> 
> 
> Error: 2 results for http://localhost:45821/about.html
> Report: https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1640813497053-27107.report.html
> 
> ❌ `categories.accessibility` failure for `minScore` assertion
> Expected >= 0.95, but found 0.94
> 
> ❌ `categories.best-practices` failure for `minScore` assertion
> Expected >= 0.95, but found 0.93
> Error: 2 results for http://localhost:45821/binder.html
> Report: https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1640813497456-16487.report.html
> 
> ❌ `categories.accessibility` failure for `minScore` assertion
> Expected >= 0.95, but found 0.85
> 
> ❌ `categories.seo` failure for `minScore` assertion
> Expected >= 0.95, but found 0.92
> Error: 1 result for http://localhost:45821/documentation.html
> Report: https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1640813498306-7357.report.html
> 
> ❌ `categories.accessibility` failure for `minScore` assertion
> Expected >= 0.95, but found 0.94

But we could go fancy and have:

image

This would require authorizing https://github.com/apps/lighthouse-ci ; it is tied to a private account of maintainer but also mentioned on https://github.com/GoogleChrome/lighthouse-ci/blob/main/docs/getting-started.md#github-status-checks.

Just an option - I don't have opinion on this one.

**Comment by choldgraf** [comment]
I'd be +1 on this personally, though also don't think we should block this PR on that if it'll require extra decision-making steps.

Is there a place where we can look at the lighthouse report itself, to quickly spot-check improvements we need to make? (e.g., similar to how the codecov reports can work?)

**Comment by krassowski** [comment]
> Is there a place where we can look at the lighthouse report itself, to quickly spot-check improvements we need to make? (e.g., similar to how the codecov reports can work?)

The links in the logs should work, for example https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1640813497053-27107.report.html for about.html highlights that the change of the links colour to orange reduced accessibility.

File: .github/workflows/validate.yml
Comment (choldgraf) [inline .github/workflows/validate.yml:82]: I had no idea there was a pytest link checker - that is cool!

File: .github/workflows/validate.yml
Comment (choldgraf) [inline .github/workflows/validate.yml:46]: Could we add a comment somewhere just saying what we mean by "validate"? e.g., are we just doing some kind of basic integrity check? And would this not be captured by lighthouse already?

**Review by choldgraf — COMMENTED** [review]
This looks great to me - only two quick comments, and one suggestion below:

Could you add a short section to the README.md file to explain this workflow? Just a short explanation of these validation steps, and short instructions for how people can view a lighthouse report. Whatever somebody might need to understand how to make the use of this helpful CI/CD infrastructure.

File: .github/workflows/validate.yml
Comment (krassowski) [reply .github/workflows/validate.yml:84]: It's homegrown: https://github.com/jupyterlab/pytest-check-links

**Comment by choldgraf** [comment]
Also just a note that I think one of the failing accessibility checks is for the new "jupyter orange" link colors that @palewire recently added in #558 :

https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1640813497053-27107.report.html#accessibility

I wonder if we could darken those a bit to improve accessibility?

File: .github/workflows/validate.yml
Comment (krassowski) [reply .github/workflows/validate.yml:46]: Some of them overlap (like `alt` meta tags for images, but others are not caught by Lighthouse).

By validate we mean run though W3C-sanctioned HTML integrity validator. In this case we use https://github.com/validator (https://validator.github.io/validator/) backend "Nu". It is featured on W3C: https://validator.w3.org/nu/ and https://validator.w3.org/docs/help.html#validation_basics describes what validation means to them.

**Comment by palewire** [comment]
Yes, I can darken it with a separate PR. Hmm. The default brand color isn't accessible on a white background? That might be a separate, larger problem.

**Comment by palewire** [comment]
Also, the fact that it was surfaced is vindication of this excellent idea for a test!

**Comment by palewire** [comment]
Patch: https://github.com/jupyter/jupyter.github.io/pull/581
Bug report: https://github.com/jupyter/design/issues/67

File: .github/workflows/validate.yml
Comment (choldgraf) [reply .github/workflows/validate.yml:46]: then maybe we could just add a comment like `# See https://validator.w3.org/docs/help.html#validation_basics for more information`?

Comment (krassowski) [reply .github/workflows/validate.yml:46]: I added link to that to README in e62e6b030436365a3c04d98c81d41a4f00a7802f

**Comment by palewire** [comment]
This is super super cool, IMHO, and potentially a model for many other publishers. You want to merge it first, or try to sort out the many linty things it raises? I'm fine with a merge.

**Comment by krassowski** [comment]
I would say it is better to merge now and iterate. @sunnyjain15699 volunteered to address the validator failures (https://github.com/jupyter/jupyter.github.io/issues/526#issuecomment-1002645447) so I did not attempt to squash them and focused on automation work, but I did fix some link issues.

**Review by palewire — APPROVED** [review]

**Comment by choldgraf** [comment]
@krassowski you ready for a merge now? Or other stuff you wanna do?

**Comment by krassowski** [comment]
Ready to merge!

**Comment by krassowski** [comment]
Ok, I extracted the follow up tasks to new issues and will merge this one in 10 minutes if there are no further comments.

**Comment by palewire** [comment]
Congratulations. This is well done. I'm already thinking about how how team at latimes.com could add something like this our build tool https://github.com/datadesk/baker-example-page-template

**Comment by choldgraf** [comment]
Then let's do it! Thanks @krassowski 🙂 and thanks @palewire for the reviews and ideas

**Pull request merged**
